pkg/arvo/lib/pull-hook.hoon

```diff
@@ -295,72 +312,310 @@
-        =^  cards  pull-hook
-          (on-leave:og path)
-        [cards this]
-      ::
-      ++  on-arvo
-        |=  [=wire =sign-arvo]
-        ^-  [(list card:agent:gall) agent:gall]
-        =^  cards  pull-hook
-          (on-arvo:og wire sign-arvo)
-        [cards this]
-      ++  on-fail
-        |=  [=term =tang]
-        ^-  [(list card:agent:gall) agent:gall]
-        =^  cards  pull-hook
-          (on-fail:og term tang)
-        [cards this]
-      ++  on-peek   
-        |=  =path
-        ^-  (unit (unit cage))
-        ?:  =(/x/dbug/state path)
-          ``noun+(slop !>(state(inner-state *vase)) on-save:og)
-        ?.  =(/x/tracking path)
-          (on-peek:og path)
-        ``noun+!>(~(key by tracking))
+      =/  versioned=?
+        ?=(%resource i.t.t.t.wire)
+      =^  [cards=(list card) hook=_pull-hook]  state
+        tr-abet:(tr-sign:(tr-abed:track-engine:hc rid) sign versioned)
+      =.  pull-hook  hook
+      [cards this]
+    ::
+    ++  on-leave
+      |=  =path
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-leave:og path)
+      [cards this]
+    ::
+    ++  on-arvo
+      |=  [=wire =sign-arvo]
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-arvo:og wire sign-arvo)
+      [cards this]
+    ::
+    ++  on-fail
+      |=  [=term =tang]
+      ^-  [(list card:agent:gall) agent:gall]
+      =^  cards  pull-hook
+        (on-fail:og term tang)
+      [cards this]
+    ::
+    ++  on-peek   
+      |=  =path
+      ^-  (unit (unit cage))
+      ?:  =(/x/dbug/state path)
+        ``noun+(slop !>(state(inner-state *vase)) on-save:og)
+      ?.  =(/x/tracking path)
+        (on-peek:og path)
+      ``noun+!>(~(key by tracking))
     --
   |_  =bowl:gall
   +*  og   ~(. pull-hook bowl)
+      io   ~(. agentio bowl)
+      pass  pass:io
+      virt  ~(. pull-hook-virt bowl)
+      ver  ~(. versioning [bowl [update-mark version min-version]:config])
+  ::
+  ++  track-engine
+    |_  [cards=(list card) rid=resource =ship =status gone=_|]
+    ::  +|  %init: state machine setup and manipulation
+    ::
+    ++  tr-core  .
+    ++  tr-abed
+      |=  r=resource
+      =/  [s=^ship sta=^status]
+        (~(got by tracking) r)
+      tr-core(rid r, ship s, status sta)
+    ::
+    ++  tr-abet
+      ^-  [[(list card) _pull-hook] _state]
+      =.  tracking
+        ?:  gone
+          (~(del by tracking) rid)
+        (~(put by tracking) rid [ship status])
+      [[(flop cards) pull-hook] state]
+    ::
+    ++  tr-emit
+      |=  =card
+      tr-core(cards [card cards])
+    ::
+    ++  tr-emis
+      |=  caz=(list card)
+      tr-core(cards (welp (flop cards) cards))
+    ::
+    ++  tr-ap-og
+      |=  ap=_^?(|.(*(quip card _pull-hook)))
+      =^  caz  pull-hook
+        (ap)
+      (tr-emis caz)
+    ::  +|  %sign: sign handling
+    ::
+    ::
+    ++  tr-sign
+      |=  [=sign:agent:gall versioned=?]
+      ?+   -.sign  !!
```

Comment:
Seems best to put tr-wack and tr-fact into a |^ here, but since tr-kick is used elsewhere seems best to keep that arm where it is.